nimbus/src/enrollment.rs

```diff
@@ -179,55 +525,117 @@ pub fn reset_enrollment(
+pub fn opt_in_with_branch(
+    db: &Database,
+    experiment_slug: &str,
+    branch: &str,
+) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
+    if !exp
+        .branches
+        .iter()
+        .map(|b| &b.slug)
+        .any(|branch_slug| branch == branch_slug)
+    {
+        // XXX - do we want specific errors for this kind of thing?
+        log::warn!(
+            "No such branch '{}' for experiment {}",
+            branch,
+            experiment_slug
+        );
+        return Ok(vec![]);
+    }
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::Enrolled {
-            reason: EnrolledReason::OptIn,
-            branch: branch.to_string(),
-        },
+        status: EnrollmentStatus::enrolled(EnrolledReason::OptIn, branch),
+    };
+    let enrollment_id = if let EnrollmentStatus::Enrolled { enrollment_id, .. } = &enrollment.status
+    {
+        enrollment_id
+    } else {
+        unreachable!(); // This is really dumb.
     };
-    let mut writer = db.write()?;
     db.get_store(StoreId::Enrollments)
         .put(&mut writer, experiment_slug, &enrollment)?;
     writer.commit()?;
-    Ok(())
+    Ok(vec![EnrollmentChangeEvent::new(
+        &enrollment.slug,
+        enrollment_id,
+        branch,
+        None,
+        EnrollmentChangeEventType::Enrollment,
+    )])
 }
 
-pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<()> {
-    // As above - check experiment exists?
+pub fn opt_out(db: &Database, experiment_slug: &str) -> Result<Vec<EnrollmentChangeEvent>> {
+    let mut writer = db.write()?;
+    let exp_store = db.get_store(StoreId::Experiments);
+    let _exp = match exp_store.get::<Experiment>(&writer, experiment_slug)? {
+        None => {
+            // XXX - do we want specific errors for this kind of thing?
+            log::warn!("No such experiment '{}'", experiment_slug);
+            return Ok(vec![]);
+        }
+        Some(e) => e,
+    };
     let enrollment = ExperimentEnrollment {
         slug: experiment_slug.to_string(),
-        status: EnrollmentStatus::OptedOut,
+        status: EnrollmentStatus::Disqualified {
+            reason: DisqualifiedReason::OptOut,
+            // TODO: We should use the branch we're enrolled in if we had an enrollment
+            // But what do we do if we did not? I assume we want to support the "opt out
+            // before enrollment ever happens use case". Should we just make "branch" optional?
+            branch: "not-specified".to_owned(),
```

Comment:
If we weren't enrolled, I think we should stay in the `NotEnrolled` state, and maybe set reason=`OptOut`. The `Disqualified` state should only be for tracking a particular branch assignment after being enrolled.

Reply:
To clarify: we should match on the current enrollment state here:

* If `Enrolled`, move to `Disqualified` with reason `OptOut` and the enrolled branch info.
* If `NotEnrolled`, stay in `NotEnrolled` but set the reason to `OptOut`.
* If `Disqualified` or `WasEnrolled`, don't change anything.